services/src/main/java/org/keycloak/vault/DefaultVaultTranscriber.java

```diff
@@ -0,0 +1,106 @@
+/*
+ * Copyright 2019 Red Hat, Inc. and/or its affiliates
+ * and other contributors as indicated by the @author tags.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.keycloak.vault;
+
+import java.lang.ref.WeakReference;
+import java.nio.ByteBuffer;
+import java.nio.CharBuffer;
+import java.nio.charset.StandardCharsets;
+import java.util.Optional;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+
+/**
+ * Default {@link VaultTranscriber} implementation that uses the configured {@link VaultProvider} to obtain raw secrets
+ * and convert them into other types. By default, the {@link VaultProvider} provides raw secrets through a {@link ByteBuffer}.
+ * This class offers methods to convert the raw secrets into other types (such as {@link VaultCharSecret} or {@link WeakReference<String>}).
+ *
+ * @see VaultRawSecret
+ * @see VaultCharSecret
+ *
+ * @author <a href="mailto:sguilhen@redhat.com">Stefan Guilhen</a>
+ */
+public class DefaultVaultTranscriber implements VaultTranscriber {
+
+    private static final Pattern pattern = Pattern.compile("^\\$\\{vault\\.(.+?)}$");
+
+    private final VaultProvider provider;
+
+    public DefaultVaultTranscriber(final VaultProvider provider) {
+        if (provider == null) {
+            throw new IllegalArgumentException("Required VaultProvider not specified - check vault configuration");
```

Comment:
In this case we perhaps should use a special vault provider instead that returns empty secrets:
```suggestion
            this.provider = new VaultProvider() {
                @Override
                public VaultRawSecret obtainSecret(String vaultSecretId) {
                    return DefaultVaultRawSecret.forBuffer(null);
                }
            }
```
That way we do not force user to set up a vault until they need it. This would also simplify migration since there would be effectively no migration needed since there would be no vault needed.

Reply:
+1 to that, will be adding the default provider.